packages/devtools_app_shared/lib/src/service/dtd_manager.dart

```diff
@@ -18,39 +23,205 @@ class DTDManager {
+  /// If the connection fails, will retry with exponential backoff up to
+  /// [maxRetries].
+  Future<void> _connectImpl(
     Uri uri, {
     void Function(Object, StackTrace?)? onError,
+    int maxRetries = 5,
+    bool disconnectBeforeConnecting = true,
   }) async {
-    await disconnect();
+    if (disconnectBeforeConnecting) {
+      await disconnect();
+    }
+    // Enable automatic reconnect on any new connection.
+    _automaticallyReconnect = true;
 
     try {
-      final connection = await DartToolingDaemon.connect(uri);
+      final connection = await _connectWithRetries(uri, maxRetries: maxRetries);
+
       _uri = uri;
       // Set this after setting the value of [_uri] so that [_uri] can be used
       // by any listeners of the [_connection] notifier.
       _connection.value = connection;
+      _connectionState.value = ConnectedDTDState();
       _log.info('Successfully connected to DTD at: $uri');
+
+      // If a connection drops (and we hadn't disabled auto-reconnect, such
+      // as by explicitly calling disconnect/dispose), we should attempt to
+      // reconnect.
+      unawaited(connection.done.then((_) async {
+        // Trigger disconnect to ensure we emit a `null` connection to
+        // listeners.
+        await disconnectImpl(disableReconnect: false);
+        if (_automaticallyReconnect) {
+          await _connectImpl(
+            uri,
+            onError: onError,
+            // We've already disconnected above, in a way that doesn't disable
+            // reconnect and does not set connection to null (allowing screens
+            // to remain visible under connection overlays).
+            disconnectBeforeConnecting: false,
+          );
+        }
+      }));
     } catch (e, st) {
       onError?.call(e, st);
     }
   }
 
+  /// Sets the Dart Tooling Daemon connection to point to [uri].
+  ///
+  /// Before connecting to [uri], if a current connection exists, then
+  /// [disconnect] is called to close it.
+  ///
+  /// If the connection fails, will retry with exponential backoff up to
+  /// [maxRetries].
+  Future<void> connect(
+    Uri uri, {
+    void Function(Object, StackTrace?)? onError,
+    int maxRetries = 5,
+  }) {
+    // On explicit connections, we capture the connect function so that we
+    // can call it again if [reconnect()] is called.
+    final connectFunc = _lastConnectFunc =
+        () => _connectImpl(uri, onError: onError, maxRetries: maxRetries);
+    return connectFunc();
+  }
+
   /// Closes and unsets the Dart Tooling Daemon connection, if one is set.
-  Future<void> disconnect() async {
-    if (_connection.value != null) {
-      await _connection.value!.close();
+  Future<void> disconnect() => disconnectImpl();
+
+  /// Closes and unsets the Dart Tooling Daemon connection, if one is set.
+  ///
+  /// Disables automatic reconnect unless [disableReconnect] is `false`.
+  @visibleForTesting
+  Future<void> disconnectImpl({bool disableReconnect = true}) async {
```

Comment:
Could we positively word this (i.e. `{bool enableReconnect = false}`)? (To avoid the double negative on line 166 above)

Reply:
Done! I went with `allowReconnect` and made it clear on the docs that this doesn't _enable_ a reconnect, it only _allows_ it (eg., if there was a dispose of explicit disconnect call that triggered the call here, it still wouldn't reconnect).